diff --git a/Android.mk b/Android.mk
index 6880ed2..2a426cc 100644
--- a/Android.mk
+++ b/Android.mk
@@ -362,6 +362,9 @@ endif
 ifneq ($(TW_CLOCK_OFFSET),)
 	LOCAL_CFLAGS += -DTW_CLOCK_OFFSET=$(TW_CLOCK_OFFSET)
 endif
+ifeq ($(TW_EXCLUDE_TWRPAPP),true)
+    LOCAL_CFLAGS += -DTW_EXCLUDE_TWRPAPP
+endif
 LOCAL_ADDITIONAL_DEPENDENCIES += \
     dump_image \
     erase_image \
diff --git a/data.cpp b/data.cpp
index 4dfbde2..1f43909 100644
--- a/data.cpp
+++ b/data.cpp
@@ -890,9 +890,13 @@ void DataManager::SetDefaultValues()
 	mConst.SetValue("tw_oem_build", "1");
 #else
 	mConst.SetValue("tw_oem_build", "0");
+  #ifdef TW_EXCLUDE_TWRPAPP
+	mConst.SetValue("tw_app_prompt", "-1");
+  #else
 	mPersist.SetValue("tw_app_prompt", "1");
 	mPersist.SetValue("tw_app_install_system", "1");
 	mData.SetValue("tw_app_install_status", "0"); // 0 = no status, 1 = not installed, 2 = already installed
+  #endif
 #endif
 
         mData.SetValue("tw_enable_adb_backup", "0");
diff --git a/gui/theme/common/landscape.xml b/gui/theme/common/landscape.xml
index 5bbfe90..c697205 100644
--- a/gui/theme/common/landscape.xml
+++ b/gui/theme/common/landscape.xml
@@ -2725,6 +2725,7 @@
 			<checkbox>
 				<placement x="%indent%" y="%row5_y%"/>
 				<text>{@reboot_install_app_prompt_install=Prompt to install TWRP app if not installed}</text>
+				<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 				<data variable="tw_app_prompt"/>
 			</checkbox>
 
diff --git a/gui/theme/common/portrait.xml b/gui/theme/common/portrait.xml
index 1176922..e338992 100644
--- a/gui/theme/common/portrait.xml
+++ b/gui/theme/common/portrait.xml
@@ -2920,6 +2920,7 @@
 			<checkbox>
 				<placement x="%indent%" y="%row5_y%"/>
 				<text>{@reboot_install_app_prompt_install=Prompt to install TWRP app if not installed}</text>
+				<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 				<data variable="tw_app_prompt"/>
 			</checkbox>
 
diff --git a/gui/theme/common/watch.xml b/gui/theme/common/watch.xml
index 83f4b63..2c32b26 100644
--- a/gui/theme/common/watch.xml
+++ b/gui/theme/common/watch.xml
@@ -3377,6 +3377,7 @@
 			<checkbox>
 				<placement x="%indent%" y="%row3_y%"/>
 				<text>{@reboot_install_app_prompt_install=Prompt to install TWRP app if not installed}</text>
+				<condition var1="tw_app_prompt" op="!=" var2="-1"/>
 				<data variable="tw_app_prompt"/>
 			</checkbox>
 
diff --git a/minui/Android.mk b/minui/Android.mk
index cb56b73..dbe8ab7 100644
--- a/minui/Android.mk
+++ b/minui/Android.mk
@@ -90,6 +90,10 @@ ifeq ($(subst ",,$(TARGET_RECOVERY_PIXEL_FORMAT)),BGRA_8888)
   LOCAL_CFLAGS += -DRECOVERY_BGRA
 endif
 
+ifeq ($(RECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER), true)
+  LOCAL_CFLAGS += -DRECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER
+endif
+
 ifneq ($(TARGET_RECOVERY_OVERSCAN_PERCENT),)
   LOCAL_CFLAGS += -DOVERSCAN_PERCENT=$(TARGET_RECOVERY_OVERSCAN_PERCENT)
 else
diff --git a/minui/graphics_fbdev.cpp b/minui/graphics_fbdev.cpp
index bb91a57..592f2cb 100644
--- a/minui/graphics_fbdev.cpp
+++ b/minui/graphics_fbdev.cpp
@@ -121,6 +121,7 @@ GRSurface* MinuiBackendFbdev::Init() {
   memset(gr_framebuffer[0].data, 0, gr_framebuffer[0].height * gr_framebuffer[0].row_bytes);
 
   /* check if we can use double buffering */
+#ifndef RECOVERY_GRAPHICS_FORCE_SINGLE_BUFFER
   if (vi.yres * fi.line_length * 2 <= fi.smem_len) {
     double_buffered = true;
 
@@ -131,6 +132,9 @@ GRSurface* MinuiBackendFbdev::Init() {
     gr_draw = gr_framebuffer + 1;
 
   } else {
+#else
+  {
+#endif
     double_buffered = false;
 
     // Without double-buffering, we allocate RAM for a buffer to
diff --git a/prebuilt/Android.mk b/prebuilt/Android.mk
index 6e9bcfe..1ee5776 100644
--- a/prebuilt/Android.mk
+++ b/prebuilt/Android.mk
@@ -449,11 +449,13 @@ ifeq ($(TW_INCLUDE_DUMLOCK), true)
 	include $(BUILD_PREBUILT)
 endif
 
-#TWRP App "placeholder"
-include $(CLEAR_VARS)
-LOCAL_MODULE := me.twrp.twrpapp.apk
-LOCAL_MODULE_TAGS := eng
-LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
-LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
-LOCAL_SRC_FILES := $(LOCAL_MODULE)
-include $(BUILD_PREBUILT)
+ifneq ($(TW_EXCLUDE_TWRPAPP),true)
+	#TWRP App "placeholder"
+	include $(CLEAR_VARS)
+	LOCAL_MODULE := me.twrp.twrpapp.apk
+	LOCAL_MODULE_TAGS := eng
+	LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
+	LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
+	LOCAL_SRC_FILES := $(LOCAL_MODULE)
+	include $(BUILD_PREBUILT)
+endif
